common/rfb/EncodeManager.cxx

```diff
@@ -997,3 +1030,125 @@ void EncodeManager::OffsetPixelBuffer::update(const PixelFormat& pf,
+  if (abs(p.x) > 32 ||
+      abs(p.y) > 32)
+      return false;
+
+  if (abs(unionArea - checkArea) > 4096)
+    return false;
+
+  return true;
+}
+
+void EncodeManager::trackRectQuality(const Rect& rect) {
+
+  const int searchArea = rect.area();
+  struct timeval now;
+  gettimeofday(&now, NULL);
+
+  for (std::list<QualityInfo*>::iterator it = qualityList.begin(); it != qualityList.end(); it++) {
+    QualityInfo * const cur = *it;
+    const int curArea = cur->rect.area();
+    const Rect unioned = cur->rect.union_boundary(rect);
+    const int unionArea = unioned.area();
+    // Is this close enough to match?
+    // e.g. ads that change parts in one frame and more in others
+    if (rect.enclosed_by(cur->rect) ||
+        cur->rect.enclosed_by(rect) ||
+        closeEnough(unioned, unionArea, cur->rect, curArea) ||
+        closeEnough(unioned, unionArea, rect, searchArea)) {
+
+        // This existing rect matched. Set it to the larger of the two,
+        // and add to its score.
+        if (searchArea > curArea)
+          cur->rect = rect;
+
+        cur->score += SCORE_INCREMENT;
+        cur->lastUpdate = now;
+        return;
+    }
+  }
+
+  // It wasn't found, add it
+  QualityInfo *info = new QualityInfo;
+  info->rect = rect;
+  info->score = 0;
+  info->lastUpdate = now;
+  qualityList.push_back(info);
+}
+
+// Returns the change-tracked quality, 0-128, where 128 is max quality
+unsigned EncodeManager::getQuality(const Rect& rect) const {
+
+  const int searchArea = rect.area();
+
+  for (std::list<QualityInfo*>::const_iterator it = qualityList.begin(); it != qualityList.end(); it++) {
+    const QualityInfo * const cur = *it;
+    const int curArea = cur->rect.area();
+    const Rect unioned = cur->rect.union_boundary(rect);
+    const int unionArea = unioned.area();
+    // Is this close enough to match?
+    // e.g. ads that change parts in one frame and more in others
+    if (rect.enclosed_by(cur->rect) ||
+        cur->rect.enclosed_by(rect) ||
+        closeEnough(unioned, unionArea, cur->rect, curArea) ||
+        closeEnough(unioned, unionArea, rect, searchArea)) {
+
+        unsigned score = cur->score;
+        if (score > 128)
+          score = 128;
+        score = 128 - score;
+
+        return score;
+    }
+  }
+
+  return 128; // Not found, this shouldn't happen - return max quality then
+}
+
+// Returns the scaled quality, 0-9, where 9 is max
+// Optionally takes bandwidth into account
+unsigned EncodeManager::scaledQuality(const Rect& rect) const {
+
+  unsigned dynamic;
+
+  dynamic = getQuality(rect);
+
+  // The tracker gives quality as 0-128. Convert to our desired range
+  dynamic *= dynamicQualityOff;
+  dynamic += 64; // Rounding
+  dynamic /= 128;
+  dynamic += dynamicQualityMin;
+
+  // Bandwidth adjustment
+  if (!Server::preferBandwidth) {
+    // Prefer quality, if there's bandwidth available, don't go below 7
+    if (curMaxUpdateSize > 2000 && dynamic < 7)
```

Comment:
These are still a bit arbitrary, so not quite what I was after when I said to look at the bandwidth. Basically the system should make a guess on what `dynamicQualifyOff` and `dynamicQualityMin` should be set to based on:

 * Available bandwidth
 * How large this update is

It knows how many pixels it needs to push, and it knows how many bytes it has for that. That gives the system a target compression ratio. What it doesn't know is the actual compression ratio it will get. For that we can start with some hard coded values and solve that issue later. Look in the JPEG encoder for some previous numbers.